Modules/posixmodule.c

```diff
@@ -1834,17 +1837,110 @@ attributes_from_dir(LPCWSTR pszFile, BY_HANDLE_FILE_INFORMATION *info, ULONG *re
+           AccessCheck to check for generic read, write, and execute
+           access. */
+        const wchar_t *fileExtension = wcsrchr(path, '.');
+        if (fileExtension) {
+            if (_wcsicmp(fileExtension, L".exe") == 0 ||
+                _wcsicmp(fileExtension, L".bat") == 0 ||
+                _wcsicmp(fileExtension, L".cmd") == 0 ||
+                _wcsicmp(fileExtension, L".com") == 0) {
+                result->st_mode |= 0111;
+            }
+        }
+    }
+}
+
+static int
+win32_xstat_get_st_dev(const wchar_t *path,
+                       struct _Py_stat_struct *result)
+{
+    const wchar_t *rootEnd;
+    WCHAR rootBuffer[MAX_PATH];
+    WCHAR *root = rootBuffer;
+    DWORD oldErrorMode;
+    DWORD vsn = 0;
+    int retval = 0;
+
+    if (result->st_dev || !path || !path[0]) {
+        return 0;
+    }
+
+    if (PathCchSkipRoot(path, &rootEnd) ||
+        wcsncpy_s(rootBuffer, MAX_PATH, path, (rootEnd - path))) {
+        /* No root for the path, so let it use the current volume */
+        root = NULL;
+    }
+
+    /* Change the thread's error mode to avoid popping up dialogs for
+       "no disk in drive" situations. */
+    if (!SetThreadErrorMode(SEM_FAILCRITICALERRORS, &oldErrorMode)) {
+        oldErrorMode = 0;
+    }
+
+    if (GetVolumeInformationW(root, NULL, 0, &vsn, NULL, NULL, NULL, 0)) {
+        result->st_dev = vsn;
+        result->st_rdev = vsn;
+    } else {
+        retval = -1;
+    }
+
+    SetThreadErrorMode(oldErrorMode, NULL);
+    return retval;
+}
+
 static int
 win32_xstat_impl(const wchar_t *path, struct _Py_stat_struct *result,
-                 BOOL traverse)
+                 BOOL traverse, BOOL fast)
 {
     HANDLE hFile;
+    FILE_STAT_BASIC_INFORMATION statInfo;
     BY_HANDLE_FILE_INFORMATION fileInfo;
     FILE_ATTRIBUTE_TAG_INFO tagInfo = { 0 };
-    DWORD fileType, error;
+    DWORD fileType, error = 0;
     BOOL isUnhandledTag = FALSE;
     int retval = 0;
 
+    /* Try the fast path first. This is an lstat equivalent, but if we
+       don't find a symlink it'll be faster to try it first */
+    if (GetFileInformationByName(path, FileStatBasicByNameInfo,
+                                 &statInfo, sizeof(statInfo))) {
+        /* Fast path succeeded. If we're not traversing or the file isn't
+           a name surrogate reparse point, we can continue */
+        if (!traverse ||
+            !(statInfo.FileAttributes & FILE_ATTRIBUTE_REPARSE_POINT &&
+              IsReparseTagNameSurrogate(statInfo.ReparseTag))) {
+            _Py_stat_basic_info_to_stat(&statInfo, result);
+            if (!fast) {
+                win32_xstat_fixup_exec_mode(path, result);
+                /* st_dev is not included in FileStatBasicByName,
+                   so we get it separately */
+                if (win32_xstat_get_st_dev(path, result) != 0) {
+                    return -1;
+                }
+            }
+            return 0;
+        }
```

Comment:
Other than placeholder reparse points (if the [placeholder compatibility mode](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-rtlsetprocessplaceholdercompatibilitymode) is set to disguise them), I assume that `GetFileInformationByName()` opens all reparse points, like `GetFileAttributesW()` does. 

If that's the case, then the handling here has to be careful. We're trying to implement POSIX-like semantics for `follow_symlinks`, for which only name-surrogate reparse points are handled like symlinks. Thus we can only use the by-name stat result if it's not a reparse point or if it's a name-surrogate reparse point and `traverse` is false. Reparse points that are not name surrogates must always be traversed.

```suggestion
        /* Fast path succeeded. If the file isn't a reparse point or if it's
           a name-surrogate reparse point and we're not traversing, then we 
           can use the fast stat. */
        if (!(statInfo.FileAttributes & FILE_ATTRIBUTE_REPARSE_POINT) ||
            !traverse && IsReparseTagNameSurrogate(statInfo.ReparseTag)) {
            _Py_stat_basic_info_to_stat(&statInfo, result);
            if (!fast) {
                win32_xstat_fixup_exec_mode(path, result);
                /* st_dev is not included in FileStatBasicByName,
                   so we get it separately */
                if (win32_xstat_get_st_dev(path, result) != 0) {
                    return -1;
                }
            }
            return 0;
        }
        // Continue to the slow path. For efficiency, set traverse to true.
        traverse = TRUE;
```

Reply:
Ah, very good catch! Thanks